repose-aggregator/components/filters/rackspace-auth-user-filter/src/main/scala/org/openrepose/filters/rackspaceauthuser/RackspaceAuthUserFilter.scala

```diff
@@ -19,84 +19,214 @@
+    val auth = xml \\ "auth"
+    // This is actually prefixed with the "RAX-AUTH:" namespace.
+    val domain = Option((auth \ "domain" \ "@name").text)
+    val possibleUsernames = List(
+      (auth \ "rsaCredentials" \ "@username").text,
+      (auth \ "apiKeyCredentials" \ "@username").text,
+      (auth \ "passwordCredentials" \ "@username").text,
+      (auth \ "@tenantId").text,
+      (auth \ "@tenantName").text
+    )
+    val usernames = possibleUsernames.filterNot(_.isEmpty)
+
+    if (usernames.isEmpty) {
+      (None, None)
+    } else {
+      domain match {
+        case Some(d) if d.nonEmpty =>
+          (Some(d), Some(getUsername(d, usernames.head)))
+        case _ =>
+          (None, Some(usernames.head))
       }
     }
   }
 
-  override def destroy(): Unit = {
-    logger.trace("Rackspace Auth User filter destroying...")
-    configurationService.unsubscribeFrom(configFileName, this.asInstanceOf[UpdateListener[_]])
-    logger.trace("Rackspace Auth User filter destroyed.")
+  val username2_0JSON: UsernameParsingFunction = { is =>
+    val json = Json.parse(is)
+    val possibleDomain = (json \ "auth" \ "RAX-AUTH:domain" \ "name").validate[String]
+
+    val domain = possibleDomain match {
+      case s: JsSuccess[String] => Some(s.get)
+      case f: JsError =>
+        None
+    }
+
+    val possibleUsernames = List(
+      (json \ "auth" \ "RAX-AUTH:rsaCredentials" \ "username").validate[String],
+      (json \ "auth" \ "passwordCredentials" \ "username").validate[String],
+      (json \ "auth" \ "RAX-KSKEY:apiKeyCredentials" \ "username").validate[String],
+      (json \ "auth" \ "tenantId").validate[String],
+      (json \ "auth" \ "tenantName").validate[String]
+    )
+
+    val usernames = possibleUsernames.map {
+      case s: JsSuccess[String] => Some(s.get)
+      case f: JsError =>
+        logger.debug(s"2.0 JSON Parsing failure: ${JsError.toFlatJson(f)}")
+        None
+    }.filterNot(_.isEmpty)
+
+    //At this point we have a prioritized list of the username parsing, where the head of the list is more
+    // important to return than the tail. If we are empty, we didn't find anything,
+    // If we've got at least one item, return just the first
+    if (usernames.isEmpty) {
+      (None, None)
+    } else {
+      domain match {
+        case Some(d) =>
+          (Some(d), Some(getUsername(d, usernames.head.get)))
+        case _ =>
+          (None, usernames.head)
+      }
+    }
   }
 
-  override def configurationUpdated(config: RackspaceAuthUserConfig): Unit = {
-    handler.set(new RackspaceAuthUserHandler(config))
-    initialized.set(true)
+  def parseUserGroupFromInputStream(inputStream: InputStream, contentType: String, sessionIds: List[String]): Option[RackspaceAuthUserGroup] = {
+    sessionIds.map(HeaderValue).sortWith(_.quality > _.quality).toStream.flatMap({ header => Option(datastore.get(s"$ddKey:${header.value}").asInstanceOf[Option[RackspaceAuthUserGroup]]) }).headOption
```

Comment:
I could move the to stream up and squeeze a little more performance out of this, but for now i waited to do it till i cared about not processing everything